nomt/src/page_cache.rs

```diff
@@ -0,0 +1,251 @@
+        let mut shared = self.shared.lock().unwrap();
+
+        // We first check in the dirty pages, since that's where we would find the most recent
+        // version of the page. If the page is not present there, we check in the prestine cache.
+        if let Some(page) = shared.dirty.get(&page_id) {
+            return PagePromise::Now(page.clone());
+        }
+        if let Some(page) = shared.prestine.get(&page_id) {
+            return PagePromise::Now(page.clone());
+        }
+        // In case none of those are present, we check if there is an inflight fetch for the page.
+        if let Some(inflight) = shared.inflight.get(&page_id) {
+            return inflight.promise();
+        }
+        // Nope, then we need to fetch the page from the store.
+        let inflight = InflightFetch::new();
+        let promise = inflight.promise();
+        shared.inflight.insert(page_id, inflight);
+        let task = {
+            let store = self.store.clone();
+            let shared = self.shared.clone();
+            move || {
+                let entry = store
+                    .load_page(page_id)
+                    .map(Arc::new)
+                    .map_or(Page::Nil, Page::Exists);
+                let mut shared = shared.lock().unwrap();
+                // Unwrap: the operation was inserted above. It is scheduled for execution only
+                // once. It may removed only in the line below. Therefore, `None` is impossible.
+                let inflight = shared.inflight.remove(&page_id).unwrap();
+                shared.prestine.insert(page_id, entry.clone());
+                inflight.complete_and_notify(entry);
+                drop(shared);
+            }
+        };
+        self.fetch_tp.execute(task);
+        promise
+    }
+
+    /// Flushes all the dirty pages into the underlying store.
+    ///
+    /// After the commit, all the dirty pages are cleared.
+    pub fn commit(&self, tx: &mut Transaction) {
+        let mut shared = self.shared.lock().unwrap();
+        for (page_id, page) in mem::take(&mut shared.dirty) {
+            shared.prestine.insert(page_id, page.clone());
+            let page_data = match page {
+                Page::Nil => None,
+                Page::Exists(data) => Some(data),
+            };
+            tx.write_page(page_id, page_data.as_deref());
+            // It doesn't seem that the page can be in the inflight fetches at this point.
+            //
+            // If the page ended up in the dirty set, then it must have been supplanted in case
+            // there was an inflight fetch for that page. After it was supplanted, the `retrieve`
+            // function would not initiate another fetch for the same page.
+            //
+            // Therefore, we don't need to divert the inflight fetches here.
+        }
+    }
+
+    /// Creates a new cursor assuming the given root node.
+    pub fn create_cursor(&self, root: Node) -> PageCacheCursor {
```

Comment:
could be cleaner to have `PageCacheCursor::new(&PageCache, root)` and then avoid any explicit dependency on cursor type in this module. Separation of concerns.

Reply:
Yes, the same idea crossed my mind, but decided to punt on it.